### PR TITLE
Cm class nns

### PR DESCRIPTION
**Pull request by chelsealuisa**
- Added vanilla SGD and dropout SGD classification networks for comparison with BNN
- Added requirements file

### CODE REVIEW

File: classification/class_task.py [hunk 1]
```diff
@@ -88,10 +88,82 @@ def evaluate(self, test_loader):
...
+        self.scheduler = torch.optim.lr_scheduler.StepLR(self.optimiser, step_size=100, gamma=0.5)
+        print('MLP Parameters: ')
+        print(f'batch size: {self.batch_size}, input shape: {model_params["input_shape"]}, hidden units: {model_params["hidden_units"]}, output shape: {model_params["classes"]}, lr: {self.lr}')
+
+    def train_step(self, train_data):
+        self.net.train()
+        for _, (x, y) in enumerate(tqdm(train_data)):
+            x, y = x.to(DEVICE), y.to(DEVICE)
+            self.net.zero_grad()
+            self.loss_info = torch.nn.functional.cross_entropy(self.net(x), y, reduction='sum')
```
Comment (tennisonliu) [inline classification/class_task.py:142]: @Eli-Persky, can you confirm whether sampling is required when performing prediction? If so, sample=True should be set in self.net(x).

File: requirements.txt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+matplotlib==3.3.3
+tensorboard==2.4.0
```
Comment (tennisonliu) [inline requirements.txt:2]: this is a nice touch :)

File: main.py [hunk 3]
```diff
@@ -112,7 +111,6 @@ def class_trainer():
         transform=transforms.ToTensor()),
         batch_size=config.batch_size, shuffle=True, drop_last=True)
 
-        
     test_ds = DataLoader(
```
Comment (tennisonliu) [inline main.py:114]: is it possible to move the DataLoader's into data_utils.py?

File: classification/class_task.py (hunk 1 above)
Comment (chelsealuisa) [reply classification/class_task.py:142]: This is the MLP model - we only need sampling when predicting with the BNN, don't we?

File: classification/class_task.py [hunk 4]
```diff
@@ -88,10 +88,82 @@ def evaluate(self, test_loader):
...
+            'batch_size': self.batch_size,
+            'hidden_units': self.hidden_units,
+            'mode': self.mode,
+            'dropout': self.dropout
+        }
+        if self.dropout:
+            self.net = MLP_Dropout(model_params).to(DEVICE)
+        else:
+            self.net = MLP(model_params).to(DEVICE)
+        self.optimiser = torch.optim.Adam(self.net.parameters(), lr=self.lr)
```
Comment (tennisonliu) [inline classification/class_task.py:132]: not too fussed about this, but if we are comparing to vanilla SGD, it would be more accurate to use SGD optimiser for training.

Comment (chelsealuisa) [reply classification/class_task.py:132]: Oh yeah, good point, will fix. Should we be using SGD for the Bayesian models then as well?

File: main.py (hunk 3 above)
Comment (chelsealuisa) [reply main.py:114]: Yes, will fix.

File: classification/class_task.py (hunk 1 above)
Comment (tennisonliu) [reply classification/class_task.py:142]: Sorry, meant to tag this line: https://github.com/tennisonliu/bayesian-neural-network/blob/cm_class_nns/classification/class_task.py#L72. which is sampling for BNN predictions.

File: classification/class_task.py (hunk 4 above)
Comment (tennisonliu) [reply classification/class_task.py:132]: Both SGD and ADAM are SGD techniques but ADAM is adaptive. I think it should be fine if we still use ADAM for Bayesian models? But for MLP, it is a bit misleading if we claim to be using SGD.

File: classification/class_task.py (hunk 1 above)
Comment (Eli-Persky) [reply classification/class_task.py:142]: I guess so but I'm not really sure what that flag actually does tbh.

Comment (tennisonliu) [reply classification/class_task.py:142]: ahh, so it would sample from the weight distributions, if the flag is not set, it will use the mean.

see this line: https://github.com/tennisonliu/bayesian-neural-network/blob/master/networks.py#L75

Comment (Eli-Persky) [reply classification/class_task.py:142]: Gotcha

File: classification/class_task.py (hunk 4 above)
Comment (chelsealuisa) [reply classification/class_task.py:132]: I'm happy to leave the BNNs with adam and change the MLPs to SGD. @Eli-Persky also happy with this?

File: classification/class_task.py (hunk 1 above)
Comment (chelsealuisa) [reply classification/class_task.py:142]: Sorry, I don't quite follow. Does sthg in the code need to change?

File: classification/class_task.py (hunk 4 above)
Comment (chelsealuisa) [reply classification/class_task.py:132]: @tennisonliu  still with the lr scheduler, or remove that also?

File: classification/class_task.py (hunk 1 above)
Comment (tennisonliu) [reply classification/class_task.py:142]: ah yep, on this line https://github.com/tennisonliu/bayesian-neural-network/blob/cm_class_nns/classification/class_task.py#L72, it would be self.net(X, sample=True)

File: classification/class_task.py (hunk 4 above)
Comment (tennisonliu) [reply classification/class_task.py:132]: the lr scheduler works with any optimiser! so we can keep as-is?

Comment (chelsealuisa) [reply classification/class_task.py:132]: is it still "vanilla SGD"?

Comment (tennisonliu) [reply classification/class_task.py:132]: yea i reckon its fine

**Review by tennisonliu — APPROVED** [review]
